project-fortis-spark/src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala

```diff
@@ -83,43 +80,28 @@ object CassandraEventsSink {
         consistencyLevel = ConsistencyLevel.ALL // Ensure write is consistent across all replicas
       )
 
-      Timer.time(Log.logDependency("sinks.cassandra", s"write.$TableEvent", _, _)) {
-        events.saveToCassandra(KeyspaceName, TableEvent, writeConf = conf)
+      Timer.time(Log.logDependency("sinks.cassandra", s"write.${Table.Events}", _, _)) {
+        events.saveToCassandra(KeyspaceName, Table.Events, writeConf = conf)
       }
     }
 
+    // TODO: push filter down to Cassandra to minimize read data volume
     def withoutDuplicates(events: RDD[Event]): RDD[Event] = {
-      val eventsRepartitioned = events.repartitionByCassandraReplica(KeyspaceName, TableEventBatches)
+      val eventsRepartitioned = events.repartitionByCassandraReplica(KeyspaceName, Table.Events)
 
-      eventsRepartitioned.joinWithCassandraTable(
+      // [new events] = [batch] - [events table]
+      eventsRepartitioned.leftJoinWithCassandraTable(
```

Comment:
Could we add a dependency timer for this so that we know how expensive it is to perform the join? This looks like a likely bottleneck that may want to be addressed in the future (e.g. using a different store for the already ingested event ids like Redis).

Reply:
Unfortunately we can't time this since it's a Spark transformation (lazy). The Spark UI should have timing information (hopefully granular enough for us to see if this is a problem). It should be possible to push the `where` clause down to Cassandra itself as well, but I wasn't quickly able to figure out how to achieve that with the Cassandra connector. Redis would likely be much faster even still.